block/stateupdate_validator.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/dymensionxyz/dymint/types"
+)
+
+// StateUpdateValidator is a validator for messages gossiped in the p2p network.
+type StateUpdateValidator struct {
+	logger       types.Logger
+	blockManager *Manager
+}
+
+// NewValidator creates a new Validator.
+func NewStateUpdateValidator(logger types.Logger, blockManager *Manager) *StateUpdateValidator {
+	return &StateUpdateValidator{
+		logger:       logger,
+		blockManager: blockManager,
+	}
+}
+
+func (v *StateUpdateValidator) ValidateStateUpdate(batch *settlement.ResultRetrieveBatch) error {
+	v.logger.Debug("validating state update", "start height", batch.StartHeight, "end height", batch.EndHeight)
+
+	err := v.validateDRS(batch.StartHeight, batch.EndHeight, batch.DRSVersion)
+	if err != nil {
+		return err
+	}
+
+	var daBlocks []*types.Block
+	var p2pBlocks []*types.Block
+
+	// load blocks for the batch height, either P2P or DA blocks
+	for height := batch.StartHeight; height <= batch.EndHeight; height++ {
+		source, err := v.blockManager.Store.LoadBlockSource(height)
+		if err != nil {
+			continue
+		}
+		block, err := v.blockManager.Store.LoadBlock(height)
+		if err != nil {
+			continue
+		}
+		if source == types.DA.String() {
+			daBlocks = append(daBlocks, block)
+		} else {
+			p2pBlocks = append(p2pBlocks, block)
+		}
+	}
+
+	if uint64(len(daBlocks)) != batch.NumBlocks {
+		daBlocks = []*types.Block{}
+		var daBatch da.ResultRetrieveBatch
+		for {
+			daBatch = v.blockManager.Retriever.RetrieveBatches(batch.MetaData.DA)
+			if daBatch.Code == da.StatusSuccess {
+				break
+			}
+		}
+		for _, batch := range daBatch.Batches {
+			daBlocks = append(daBlocks, batch.Blocks...)
+		}
+	}
+
+	// validate DA blocks against the state update
+	err = v.ValidateDaBlocks(batch, daBlocks)
+	if err != nil {
+		return err
+	}
+
+	// compare the batch blocks with the blocks applied from P2P
+	err = v.ValidateP2PBlocks(daBlocks, p2pBlocks)
+	if err != nil {
+		return err
+	}
+
+	// update the last validated height to the batch last block height
+	v.blockManager.State.SetLastValidatedHeight(batch.EndHeight)
+	return nil
+}
+
+func (v *StateUpdateValidator) ValidateP2PBlocks(daBlocks []*types.Block, p2pBlocks []*types.Block) error {
+	// nothing to compare
+	if len(p2pBlocks) == 0 {
+		return nil
+	}
+
+	// iterate over daBlocks and compare hashes if there block is also in p2pBlocks
+	i := 0
+	for _, daBlock := range daBlocks {
+
+		if p2pBlocks[i].Header.Height != daBlock.Header.Height {
+			break
+		}
+		p2pBlockHash, err := blockHash(p2pBlocks[i])
+		if err != nil {
+			return err
+		}
+		daBlockHash, err := blockHash(daBlock)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(p2pBlockHash, daBlockHash) {
+			return types.NewErrStateUpdateDoubleSigningFraud(daBlock.Header.Height)
+		}
+		i++
+		if i == len(p2pBlocks) {
+			break
+		}
+	}
+	return nil
+}
+
+func (v *StateUpdateValidator) ValidateDaBlocks(slBatch *settlement.ResultRetrieveBatch, daBlocks []*types.Block) error {
```

Comment:
we verify all block descriptors against all blocks in DA. these blocks could be the same that are validated before applied (because we synced from DA), but if not, we are actually comparing the specific DA blocks are the same as the ones applied (from p2p) in ValidateP2PBlocks, and therefore all applied blocks are directly (or indirectly) validated against bds. although not 100% sure this is what you asking here....  in any case i improved doc to make it clear.

Reply:
I'm just wondering is it possible a node does a fraud only in DA and never produces a gossip block
so in this way we will never notice it?

but I guess we can assume that even if the malicious sequencer does not gossip blocks directly, someone else will get them from DA and gossip them?

Reply:
not sure what you mean. All blocks in DA are verified (applied or not). the validator downloads all  DA blocks and verifies them against SL batch (at block descriptor level).  all applied blocks are verified at block level (implemented block production detection misbehavior PR). if blocks are not applied from DA are compared and verified that are exactly the same block applied, so indirectly all DA blocks are validated from block descriptors and at block level.